### PR TITLE
bpo-43973: Use Py_TPFLAGS_IMMUTABLETYPE to check for class assignments

### PR DESCRIPTION
**Pull request by erlend-aasland**
See also [bpo-43908](https://bugs.python.org/issue43908)

https://bugs.python.org/issue43973

### CODE REVIEW

**Comment by erlend-aasland** [comment]
cc. @vstinner

See [msg392291](https://bugs.python.org/issue43973#msg392291): Can we get rid of the huge comment preceding the if statement now? Or at least reduce it considerably.

**Comment by erlend-aasland** [comment]
AFAICS, this does not require a news item, but I might be wrong.

**Comment by shreyanavigyan** [comment]
Doesn't this also require a change?

https://github.com/python/cpython/blob/e047239eafefe8b19725efffe7756443495cf78b/Objects/typeobject.c#L4679-L4690

Suggested :-
```
 if (compatible_for_assignment(oldto, newto, "__class__")) { 
     if (!(newto->tp_flags & Py_TPFLAGS_IMMUTABLETYPE)) { 
         Py_INCREF(newto); 
     } 
     Py_SET_TYPE(self, newto); 
     if (!(oldto->tp_flags & Py_TPFLAGS_IMMUTABLETYPE))
         Py_DECREF(oldto); 
     return 0; 
 } 
 else { 
     return -1; 
 } 
```

**Comment by erlend-aasland** [comment]
> Doesn't this also require a change?

AFAICT, no. They adjust the ref. count based on if this is a heap type or not.

File: Objects/typeobject.c [hunk 1]
```diff
@@ -4668,10 +4668,10 @@ object_set_class(PyObject *self, PyObject *value, void *closure)
     */
     if (!(PyType_IsSubtype(newto, &PyModule_Type) &&
           PyType_IsSubtype(oldto, &PyModule_Type)) &&
-        (!(newto->tp_flags & Py_TPFLAGS_HEAPTYPE) ||
-         !(oldto->tp_flags & Py_TPFLAGS_HEAPTYPE))) {
+        ((newto->tp_flags & Py_TPFLAGS_IMMUTABLETYPE) ||
+         (oldto->tp_flags & Py_TPFLAGS_IMMUTABLETYPE))) {
         PyErr_Format(PyExc_TypeError,
-                     "__class__ assignment only supported for heap types "
+                     "__class__ assignment only supported for mutable types "
```
Comment (vstinner) [inline Objects/typeobject.c:4743]: it's also supported on modules.

```
vstinner@apu$ python3
Python 3.9.4 (default, Apr  6 2021, 00:00:00) 
>>> import sys
>>> import os
>>> ModuleType=type(sys)
>>> class MyModuleType(ModuleType): pass
... 
>>> type(sys)
<class 'module'>
>>> sys.__class__=MyModuleType
>>> type(sys)
<class '__main__.MyModuleType'>
>>> type(os)
<class 'module'>
```

I'm not sure why there is a special case for module objects.

Comment (erlend-aasland) [reply Objects/typeobject.c:4743]: > it's also supported on modules.

Yes, and the error text still includes that information :) 

> I'm not sure why there is a special case for module objects.

See the comment above the if :) Also, see the other bpo referenced in this issue.

Comment (vstinner) [reply Objects/typeobject.c:4743]: I missed "or ModuleType subclasses" line... Ignore my comment :-)

**Comment by erlend-aasland** [comment]
Thanks for reviewing, @vstinner !

**Comment by vstinner** [comment]
Merged, thanks for the fix.

**Comment by pitrou** [comment]
The comment above the check should have been fixed to reflect the new semantics, no?

Also, is the explicit check for `PyModule_Type` still needed?

**Comment by erlend-aasland** [comment]
> The comment above the check should have been fixed to reflect the new semantics, no?

Yes, and I believe it can be considerably reduced.

> Also, is the explicit check for `PyModule_Type` still needed?

I don't know. See bpo-24912.

**Pull request merged**